### PR TITLE
Refactoring book details

### PR DESCRIPTION
**Pull request by Sspadea**
## Summary ##

**Trello card**: https://trello.com/c/TM8GAP6P/20-n-71-extraer-componentes-en-controllers-separados

- Refactor details view to components

## Screenshots ##

Captura de Pantalla 2022-04-07 a la(s) 14 58 10

### CODE REVIEW

File: WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        private func setBookDetails() {
+            let book = bookDetailsCardViewModel.book
+            bookDetailsCardView.titleLabel.text = book.title
+            bookDetailsCardView.statusLabel.text = book.status.localized()
+            bookDetailsCardView.authorLabel.text = book.author
+            bookDetailsCardView.yearLabel.text = book.year
+            bookDetailsCardView.genreLabel.text = book.genre
+            bookDetailsCardView.bookCoverImage.image = .cover
+           
+            if book.status == "Available" {
```
Comment (dfmm472) [inline WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift]: I think Available should be a constant

File: WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                bookDetailsCardView.statusLabel.textColor = .available
+            } else {
+                bookDetailsCardView.statusLabel.textColor = .unavailable
+            }
+        }
+        
+        private func rentBook() {
+            if bookDetailsCardView.statusLabel.text == "Unavailable".localized() {
```
Comment (dfmm472) [inline WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift]: `bookDetailsCardView.statusLabel.text == "Unavailable".localized()
`I think that the language of the cell phone can affect this comparison, the value should be in a constant

File: WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                showAlertBookUnavailble(title: "ALERT_TITLE".localized(), message: "ALERT_MESSAGE".localized())
+            } else {
+                loadRent()
+            }
+        }
+        
+        private func loadRent() {
+            bookDetailsCardViewModel.rentBook { [weak self] in
+                self?.bookDetailsCardView.statusLabel.text = "Unavailable".localized()
```
Comment (dfmm472) [inline WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift:62]: I think that the language of the cell phone can affect this comparison, the value should be in a constant

File: WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                self?.bookDetailsCardView.statusLabel.textColor = .unavailable
+            }
+        }
+
+        func showAlertBookUnavailble(title: String, message: String) {
+            let alert = UIAlertController(title: title, message: message, preferredStyle: UIAlertController.Style.alert)
+            alert.addAction(UIAlertAction(title: "Ok", style: UIAlertAction.Style.default, handler: nil))
```
Comment (dfmm472) [inline WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift]: The button text can be in a localizable

File: WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    
+    func rentBook(action: @escaping () -> Void) {
+        let onSuccess = { (_: RentResponse) in
+            action()
+        }
+        let onError = { error in print(error) }
+        rentRepository.rentBook( onSuccess: onSuccess,
+                                 onError: onError,
+                                 parameters: getParametersRent(),
+                                 userID: 10)
```
Comment (dfmm472) [inline WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewModel.swift]: The value of the id can be in a constant
`let userID = 10`

File: WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift (hunk 3 above)
Comment (Sspadea) [reply WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift:62]: This is not a comparison I am modifying the text.

Comment (dfmm472) [reply WBooks/BookDetails/BookDetailsCard/BookDetailsCardViewController.swift:62]: I'm sorry, the eagerness played a trick on me ahaha

**Review by dfmm472 — APPROVED** [review]

File: WBooks/BookDetails/BookDetailsViewModel.swift [hunk 6]
```diff
@@ -7,59 +7,24 @@
...
-    private var commentRepository = CommentRepository()
-    private var userRepository = UserRepository()
-    private(set) var user: User?
+protocol BookDetailsViewModelProtocol {
+    func bookDetailsCardViewModel() -> BookDetailsCardViewModel
+    func bookCommentsCardViewModel() -> BookCommentsCardViewModel
+    var book: Book { get set }
+}
+
+class BookDetailsViewModel: BookDetailsViewModelProtocol {
```
Comment (sleonardo) [inline WBooks/BookDetails/BookDetailsViewModel.swift:16]: Add public class

File: WBooks/BookDetails/BookCommentsCard/BookCommentsCardViewController.swift [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return bookCommentsCardViewModel.comments.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath)
+                as? BookCommentsCell else { return UITableViewCell() }
+        let comment = bookCommentsCardViewModel.comments[indexPath.row]
+        bookCommentsCardViewModel.getUser(id: comment.userID) { [weak self] in
+            cell.usernameLabel.text = self?.bookCommentsCardViewModel.user?.username
```
Comment (sleonardo) [inline WBooks/BookDetails/BookCommentsCard/BookCommentsCardViewController.swift]: Pass this logic to the UIView. When we use the MVVM architecture, we should consider having a
UIView that has all the references and handles the logic of the view

**Review by sleonardo — APPROVED** [review]

**Pull request merged**